src/mtev_cluster.c

```diff
@@ -223,35 +253,80 @@ mtev_cluster_check_timeouts() {
 
 static int 
 mtev_cluster_info_compose(void *payload, int len, void *c) {
+  hb_payload_t *hb_payload;
+
   mtev_cluster_check_timeouts();
 
   MEMWRITE_DECL(payload, len);
   u_int64_t packed_time;
-  u_int64_t seq;
+  u_int64_t my_seq;
   mtev_cluster_t *cluster = c;
+  mtev_hash_iter payload_iter = MTEV_HASH_ITER_ZERO;
+  mtev_hash_iter payload_iter2 = MTEV_HASH_ITER_ZERO;
+  void *vdata;
+  const char *key;
+  int klen;
+
   /* payload about the local cluster goes here */
   /* The header is always the same: <uuid:16><boottime_sec:5><boottiime_usec:3><seq:8> */
   MEMWRITE(my_cluster_id, UUID_SIZE);
   packed_time = (my_boot_time.tv_sec & 0x000000ffffffffffULL) << 24; /* 5 bytes << 3 bytes */
   packed_time |= (my_boot_time.tv_usec & 0xffffff); /* 3 bytes */
   packed_time = htonll(packed_time);
   MEMWRITE(&packed_time, sizeof(packed_time));
-  seq = htonll(cluster->config_seq);
-  MEMWRITE(&seq, sizeof(seq));
+  my_seq = htonll(cluster->config_seq);
+  MEMWRITE(&my_seq, sizeof(my_seq));
+
+  assert(HEART_BEAT_HDR_LEN == MEMWRITE_WRITTEN);
+
+  if(mtev_hash_size(&cluster->hb_payloads) != 0) {
+    // write header (pointer table)
+    while(mtev_hash_next(&cluster->hb_payloads, &payload_iter, &key, &klen, &vdata)) {
+      hb_payload = vdata;
+      MEMWRITE(&hb_payload->app_id, 1);
+      MEMWRITE(&hb_payload->key, 1);
+      MEMWRITE(&hb_payload->data_len, 1);
+    }
+    // write payload
+    while(mtev_hash_next(&cluster->hb_payloads, &payload_iter2, &key, &klen, &vdata)) {
+      hb_payload = vdata;
+      MEMWRITE(hb_payload->data, hb_payload->data_len);
```

Comment:
This is confusing... you write out all the headers first, then all the payloads?  This is atypical serialization.  Shouldn't it be one iter loop that writes (1,1,1,data_len) bytes, followed by another of the same?  How would you know where all the sequential headers stop and the data starts?